core/error_codes.rs

```diff
@@ -0,0 +1,212 @@
+use crate::error::AnyError;
+
+pub(crate) fn get_error_code(err: &AnyError) -> Option<&'static str> {
+  err
+    .downcast_ref::<std::io::Error>()
+    .map(|e| match e.raw_os_error() {
+      Some(code) => get_os_error_code(code),
+      None => get_io_error_code(e),
```

Comment:
If there is no OS error code, I don't think node would map to any of the `code`s that are defined in this PR.

It assigns other `code`s though (this one from OpenSSL for example):

```
piscisaureus@dumbo:~/d/deno 🦕 node
Welcome to Node.js v17.0.1.
Type ".help" for more information.
> void tls.connect(80, "deno.land").on("error", e => console.log(e.code))
undefined
> ERR_SSL_WRONG_VERSION_NUMBER
```

Reply:
IOW we should probably remove this branch and assign no code at all.

Reply:
This branch is required, `std::io::Error` is really just an enum of:
```rs
enum Repr {
    Os(i32),
    Simple(ErrorKind),
    // &str is a fat pointer, but &&str is a thin pointer.
    SimpleMessage(ErrorKind, &'static &'static str),
    Custom(Box<Custom>),
}
```

Without this `ErrorKind` classification, `Simple` & `SimpleMessage` errors don't get codes and it looks like rust's `std` sometimes creates these "synthetic" errors instead of passing through or emulating OS ones. (e.g: when you fail an operation based on stat info instead of doing the actual syscall)

I tried removing it and it breaks node apps like `yarn` which need these error codes during their config probing phase.